File: 02.hs
main = readFile "02-input.txt" >>= putStrLn . show . (runProgram 0) . restore . (map read) . split

valid :: Char -> Bool
valid c = c >= '0' && c <= '9'

split :: String -> [String]
split [] = []
split s =
  let
    s'     = dropWhile (not . valid) s
    first  = takeWhile valid s'
    second = dropWhile valid s'
  in
    if first == [] then [] else first : split second

setValue value 0 (x:xs) = value : xs
setValue value i (x:xs) = x : setValue value (i - 1) xs

restore :: [Int] -> [Int]
restore p = setValue 2 2 (setValue 12 1 p)

getOp :: Int -> Int -> Int -> Int
getOp 1 = (+)
getOp 2 = (*)

runProgram :: Int -> [Int] -> Int
runProgram i p =
  let
    opcode = p !! i
  in
    if opcode == 99 then
      p !! 0
    else
      let
        op1 = p !! (p !! (i + 1))
        op2 = p !! (p !! (i + 2))
        value = getOp opcode op1 op2
        dst = p !! (i + 3)
      in
        runProgram (i + 4) (setValue value dst p)

File: 02.rs
fn main() {
  let file = std::fs::read_to_string("02-input.txt").unwrap();
  let p = file.trim().split(',').map(str::parse).map(Result::unwrap).collect();
  find_inputs(p);
}

fn restore(mut p: Vec<i32>, noun: i32, verb: i32) -> Vec<i32> {
  p[1] = noun;
  p[2] = verb;
  p
}

fn get_op(opcode: i32) -> fn(i32, i32) -> i32 {
  use std::ops::{ Add, Mul };
  match opcode {
    1 => i32::add,
    2 => i32::mul,
    _ => unreachable!(),
  }
}

fn run_program(mut p: Vec<i32>) -> i32 {
  let mut i = 0;
  while p[i] != 99 {
    let opcode = p[i];
    let arg1 = p[p[i+1] as usize];
    let arg2 = p[p[i+2] as usize];
    let dst = p[i+3] as usize;
    p[dst] = get_op(opcode)(arg1, arg2);
    i += 4;
  }
  p[0]
}

fn find_inputs(p: Vec<i32>) {
  for noun in 0..=99 {
    for verb in 0..=99 {
      let result = run_program(restore(p.clone(), noun, verb));
      if result == 19690720 {
        println!("noun = {}", noun);
        println!("verb = {}", verb);
        println!("100 * noun + verb = {}", 100 * noun + verb);
        return;
      }
    }
  }
}
